File: sorting/MergeSort.java
import java.util.*;

public class MergeSort{
    
    public int[] intToArray(ArrayList<Integer> array){
        int[] result = new int[array.size()];
        for (int i = 0; i < array.size(); i++){
            result[i] = array.get(i);
        }
        return result;
    }

    public int[] slice(int[] array, int start, int end){
        ArrayList<Integer> al = new ArrayList<Integer>();
        for (int i = start; i < end; i++){
            al.add(array[i]);
        }
        return intToArray(al);
    }

    public int[] slice(int[] array, int start){
        ArrayList<Integer> al = new ArrayList<Integer>();
        for (int i = start; i < array.length; i++){
            al.add(array[i]);
        }
        return intToArray(al);
    }

    public int[] merge(int[] array1, int[] array2){
        ArrayList<Integer> al = new ArrayList<>();
        int i = 0;
        int j = 0;
        while ((i < array1.length) && (j < array2.length)){
            if (array1[i] < array2[j]){
                al.add(array1[i]);
                i++;
            }
            else {
                al.add(array2[j]);
                j++;
            }
        }
        while (i < array1.length){
            al.add(array1[i]);
            i++;
        }
        while (j < array2.length){
            al.add(array2[j]);
            j++;
        }
        return intToArray(al);
    }

    public int[] mergeSort(int[] array){
        if (array.length <= 1) return array;
        int mid = array.length/2;
        int[] left = mergeSort(slice(array, 0, mid));
        int[] right = mergeSort(slice(array, mid));
        return merge(left, right);        
    }

    public static void main(String[] args){
        MergeSort srt = new MergeSort();
        int[] ar1 = {78, 45, 123, 67, 11, 3, 5, 1};
        //int[] ar2 = {1, 3, 7, 9};
        System.out.println(Arrays.toString(srt.mergeSort(ar1)));
    }
}

File: sorting/merge_sort.py
def merge(ar1, ar2):
    result = []
    i = j = 0
    while (i < len(ar1) and (j < len(ar2))):
        if ar1[i] < ar2[j]:
            result.append(ar1[i])
            i += 1
        else:
            result.append(ar2[j])
            j += 1
    while (i < len(ar1)):
        result.append(ar1[i])
        i += 1
    while (j < len(ar2)):
        result.append(ar2[j])
        j += 1
    return result

def merge_sort(array):
    if len(array) <= 1:
        return array
    mid = int(len(array)/2)
    left = merge_sort(array[0:mid])
    right = merge_sort(array[mid:])
    return merge(left, right)

#print(merge([1, 3, 5, 7], [2, 4, 6, 8, 10]))
print(merge_sort([78, 34, 2, 67, 33, 21, 1]))